solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpSolrClientJdkImpl.java

```diff
@@ -0,0 +1,443 @@
+      queryParams.add(calculateQueryParams(solrRequest.getQueryParams(), requestParams));
+      return new PreparePostPutRequestResult(
+          queryParams, HttpRequest.BodyPublishers.ofString(requestParams.toString()));
+    } else {
+      return new PreparePostPutRequestResult(queryParams, HttpRequest.BodyPublishers.noBody());
+    }
+  }
+
+  private void decorateRequest(HttpRequest.Builder reqb, SolrRequest<?> solrRequest) {
+    if (requestTimeoutMillis > 0) {
+      reqb.timeout(Duration.of(requestTimeoutMillis, ChronoUnit.MILLIS));
+    } else {
+      reqb.timeout(Duration.of(idleTimeoutMillis, ChronoUnit.MILLIS));
+    }
+    reqb.header("User-Agent", USER_AGENT);
+    setBasicAuthHeader(solrRequest, reqb);
+    Map<String, String> headers = solrRequest.getHeaders();
+    if (headers != null) {
+      for (Map.Entry<String, String> entry : headers.entrySet()) {
+        reqb.header(entry.getKey(), entry.getValue());
+      }
+    }
+  }
+
+  private void setBasicAuthHeader(SolrRequest<?> solrRequest, HttpRequest.Builder reqb) {
+    if (solrRequest.getBasicAuthUser() != null && solrRequest.getBasicAuthPassword() != null) {
+      String encoded =
+          basicAuthCredentialsToAuthorizationString(
+              solrRequest.getBasicAuthUser(), solrRequest.getBasicAuthPassword());
+      reqb.header("Authorization", encoded);
+    } else if (basicAuthAuthorizationStr != null) {
+      reqb.header("Authorization", basicAuthAuthorizationStr);
+    }
+  }
+
+  private static final Pattern MIME_TYPE_PATTERN = Pattern.compile("^(.*?)(?:;| |$)");
+
+  private String contentTypeToMimeType(String contentType) {
+    Matcher mimeTypeMatcher = MIME_TYPE_PATTERN.matcher(contentType);
+    return mimeTypeMatcher.find() ? mimeTypeMatcher.group(1) : null;
+  }
+
+  private static final Pattern CHARSET_PATTERN = Pattern.compile("(?i)^.*charset=(.*)?(?:;| |$)");
+
+  private String contentTypeToEncoding(String contentType) {
+    Matcher encodingMatcher = CHARSET_PATTERN.matcher(contentType);
+    return encodingMatcher.find() ? encodingMatcher.group(1) : null;
+  }
+
+  private NamedList<Object> processErrorsAndResponse(
+      SolrRequest<?> solrRequest, ResponseParser parser, HttpResponse<InputStream> resp, String url)
+      throws SolrServerException {
+    String contentType = resp.headers().firstValue("Content-Type").orElse(null);
+    contentType = contentType == null ? "" : contentType;
+    String mimeType = contentTypeToMimeType(contentType);
+    String encoding = contentTypeToEncoding(contentType);
+    String method = resp.request() == null ? null : resp.request().method();
+    int status = resp.statusCode();
+    String reason = "" + status;
+    InputStream is = resp.body();
+    return processErrorsAndResponse(
+        status, reason, method, parser, is, mimeType, encoding, isV2ApiRequest(solrRequest), url);
+  }
+
+  @Override
+  public void close() throws IOException {
```

Comment:
Don't we just need to do instanceof here and call it?

Reply:
Wouldn't the user need to start their application with an "--adds-opens" in order to be able to get into the implementation class?  If there is a way to do this without placing such burdens on users, then I'd agree.  I think the unit test at least demonstrates that everything is cleaned up after this falls out of scope, after a full gc.  That is the best we can do until we require Jdk21, no?